src/components/ProjectSearch/index.js

```diff
@@ -20,84 +23,256 @@ import {
 
+type Match = {
+  sourceId: string,
+  line: number,
+  column: number,
+  match: string,
+  value: string,
+  text: string
+};
+type Result = {
+  filepath: string,
+  matches: Array<Match>,
+  sourceId: string
+};
+
+type State = {
+  inputValue: string,
+  inputFocused: boolean
+};
+
 type Props = {
   sources: Object,
-  results: Object,
-  textSearchQuery: string,
-  setActiveSearch: Function,
-  closeActiveSearch: Function,
-  closeProjectSearch: Function,
-  searchSources: Function,
-  activeSearch: string,
-  selectLocation: Function,
-  status: string
+  query: string,
+  results: List<Result>,
+  status: StatusType,
+  activeSearch: ActiveSearchType,
+  setActiveSearch: (activeSearch?: ActiveSearchType) => void,
+  closeProjectSearch: () => void,
+  searchSources: (query: string) => void,
+  selectLocation: (location: Location, tabIndex?: string) => void
 };
 
-class ProjectSearch extends Component<Props> {
-  onEscape: Function;
-  close: Function;
-  toggleProjectTextSearch: Function;
-
-  constructor(props) {
+export class ProjectSearch extends Component<Props, State> {
+  focusedItem: ?{
+    setExpanded?: any,
+    file?: any,
+    expanded?: any,
+    match?: Match
+  };
+  constructor(props: Props) {
     super(props);
-    this.toggleProjectTextSearch = this.toggleProjectTextSearch.bind(this);
+    this.state = {
+      inputValue: this.props.query || "",
+      inputFocused: false
+    };
   }
 
   componentDidMount() {
-    const shortcuts = this.context.shortcuts;
+    const { shortcuts } = this.context;
 
     shortcuts.on(
       L10N.getStr("projectTextSearch.key"),
       this.toggleProjectTextSearch
     );
+    shortcuts.on("Enter", this.onEnterPress);
   }
 
   componentWillUnmount() {
-    const shortcuts = this.context.shortcuts;
+    const { shortcuts } = this.context;
     shortcuts.off(
       L10N.getStr("projectTextSearch.key"),
       this.toggleProjectTextSearch
     );
+    shortcuts.off("Enter", this.onEnterPress);
   }
 
-  toggleProjectTextSearch(key, e) {
-    const { closeActiveSearch, setActiveSearch } = this.props;
+  toggleProjectTextSearch = (key: string, e: KeyboardEvent) => {
+    const { closeProjectSearch, setActiveSearch } = this.props;
     if (e) {
       e.preventDefault();
     }
 
     if (this.isProjectSearchEnabled()) {
-      return closeActiveSearch();
+      return closeProjectSearch();
     }
     return setActiveSearch("project");
+  };
+
+  isProjectSearchEnabled = () => this.props.activeSearch === "project";
+
+  selectMatchItem = (matchItem: Match) => {
+    this.props.selectLocation({ ...matchItem });
+  };
+
+  getResults = (): Result[] => {
+    const { results } = this.props;
+    return results
+      .toJS()
+      .filter(result => result.filepath && result.matches.length > 0);
+  };
+
+  getResultCount = () => {
+    const results = this.getResults();
+    return results.reduce(
+      (count, file) => count + (file.matches ? file.matches.length : 0),
```

Comment:
out of curiosity, how can a file not have `matches`?

Reply:
I don't know if it ever doesn't have matches. I had lifted it. 🙁

Reply:
Checked we don't need this ternary. Fixing.